File: src/web/Utils/ControllerBaseExtensions.cs
using System.Threading.Tasks;
using core.fs;
using core.fs.Adapters.CSV;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Primitives;
using Microsoft.FSharp.Core;

namespace web.Utils
{
    public static class ControllerBaseExtensions
    {
        public static ActionResult GenerateExport(
            this ControllerBase controller,
            ExportResponse response)
        {
            controller.HttpContext.Response.Headers.ContentDisposition = new StringValues($"attachment; filename={response.Filename}");

            return new ContentResult
            {
                Content = response.Content,
                ContentType = response.ContentType
            };
        }
        public static async Task<ActionResult> GenerateExport(
            this ControllerBase controller,
            Task<FSharpResult<ExportResponse,ServiceError>> responseTask)
            {
                var response = await responseTask;

                if (response.IsError)
                {
                    return controller.Error(response.ErrorValue.Message);
                }

                return controller.GenerateExport(response.ResultValue);
            }

        public static ActionResult Error(
            this ControllerBase controller,
            string error)
        {
            
            var dict = new Microsoft.AspNetCore.Mvc.ModelBinding.ModelStateDictionary();
            dict.AddModelError("error", error);
            return controller.BadRequest(dict);
        }
        
        public static async Task<ActionResult> OkOrError(
            this ControllerBase controller,
            Task<FSharpResult<Unit,ServiceError>> r)
        {
            return controller.OkOrError(await r);
        }
        
        public static async Task<ActionResult> OkOrError<T>(
            this ControllerBase controller,
            Task<FSharpResult<T,ServiceError>> r)
        {
            return controller.OkOrError(await r);
        }
        
        public static ActionResult OkOrError(
            this ControllerBase controller,
            FSharpResult<Unit,ServiceError> r)
        {
            if (r.IsError)
            {
                return controller.Error(r.ErrorValue.Message);
            }

            return controller.Ok();
        }
        
        public static ActionResult OkOrError<T>(
            this ControllerBase controller,
            FSharpResult<T,ServiceError> r)
        {
            if (r.IsOk == false)
            {
                return controller.Error(r.ErrorValue.Message);
            }

            return controller.Ok(r.ResultValue);
        }
    }
}

File: src/web.fs/Utils/ControllerBaseExtensions.fs
namespace web.Utils

open System.Threading.Tasks
open core.fs
open core.fs.Adapters.CSV
open Microsoft.AspNetCore.Mvc
open Microsoft.Extensions.Primitives
open Microsoft.FSharp.Core

[<AutoOpen>]
module ControllerBaseExtensions =
    
    type ControllerBase with
        
        member this.GenerateExport(response: ExportResponse) =
            this.HttpContext.Response.Headers.ContentDisposition <- StringValues($"attachment; filename={response.Filename}")
            
            ContentResult(
                Content = response.Content,
                ContentType = response.ContentType
            ) :> ActionResult
        
        member this.GenerateExport(responseTask: Task<Result<ExportResponse, ServiceError>>) = task {
            let! response = responseTask
            
            match response with
            | Error error -> return this.Error(error)
            | Ok exportResponse -> return this.GenerateExport(exportResponse)
        }
        
        member this.Error(error: ServiceError) =
            let dict = Microsoft.AspNetCore.Mvc.ModelBinding.ModelStateDictionary()
            dict.AddModelError("error", error.Message)
            this.BadRequest(dict) :> ActionResult
        
        member this.OkOrError(r: Task<Result<unit, ServiceError>>) = task {
            let! result = r
            return this.OkOrError(result)
        }
        
        member this.OkOrError(r: Task<Result<'T, ServiceError>>) = task {
            let! result = r
            return this.OkOrError(result)
        }
        
        member this.OkOrError(r: Result<unit, ServiceError>) =
            match r with
            | Error error -> this.Error(error)
            | Ok () -> this.Ok() :> ActionResult
        
        member this.OkOrError(r: Result<'T, ServiceError>) =
            match r with
            | Error error -> this.Error(error)
            | Ok value -> this.Ok(value) :> ActionResult
